File: leetcode/remove_nth_node_from_end_of_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
  ListNode *
  removeNthFromEnd(ListNode *head, int n) {
    ListNode *nth = head, *prev = NULL, *iter = head;
    for (int i = 1; i < n; ++i)
      iter = iter->next;
    while (iter->next != NULL) {
      prev = nth;
      nth = nth->next;
      iter = iter->next;
    }

    if (prev) prev->next = nth->next;
    if (nth == head) return nth->next;
    return head;
  }
};

File: leetcode/remove_nth_node_from_end_of_list.py
from list import List, Node

def removeNthFromEnd(lst, n):
    cur = lst.head
    for x in xrange(n-1): cur = cur.next
    nth = lst.head
    n_1th = None
    while (cur != lst.tail):
        cur = cur.next
        n_1th = nth
        nth = nth.next
    n_1th.next = nth.next
    if nth == lst.tail:
        lst.tail = n_1th
    del nth

lst = List([1,2,3,4,5])
removeNthFromEnd(lst, 2)
print lst

